clients/drcachesim/tools/instr_decode_cache.h

```diff
@@ -0,0 +1,172 @@
+                    instr_t *instr) = 0;
+};
+
+/**
+ * Decode info including the full decoded instr_t. This should be used with a
+ * \p instr_decode_cache_t constructed with \p persist_decoded_instrs_ set to
+ * true.
+ */
+class instr_decode_info_t : public decode_info_base_t {
+public:
+    ~instr_decode_info_t();
+    void
+    set_decode_info(void *dcontext,
+                    const dynamorio::drmemtrace::_memref_instr_t &memref_instr,
+                    instr_t *instr) override;
+
+    instr_t *instr_ = nullptr;
+
+private:
+    void *dcontext_ = nullptr;
+};
+
+/**
+ * A cache to store decode info for instructions per observed app pc. The template arg
+ * DecodeInfo is a class derived from \p decode_info_base_t which implements the
+ * set_decode_info() function that derives the required decode info from an \p instr_t
+ * object provided by \p instr_decode_cache_t. This class handles the heavy-lifting of
+ * actually producing the decoded \p instr_t. The decoded \p instr_t may be made to
+ * persist beyond the set_decode_info() calls by constructing the
+ * \p instr_decode_cache_t objects with \p persist_decoded_instrs_ set to true.
+ */
+template <class DecodeInfo> class instr_decode_cache_t {
+    static_assert(std::is_base_of<decode_info_base_t, DecodeInfo>::value,
+                  "DecodeInfo not derived from decode_info_base_t");
+
+public:
+    instr_decode_cache_t(void *dcontext, bool persist_decoded_instrs)
+        : dcontext_(dcontext)
+        , persist_decoded_instrs_(persist_decoded_instrs) {};
+
+    /**
+     * Returns a pointer to the DecodeInfo available for the instruction at \p pc.
+     * Returns nullptr if no instruction is known at that \t pc. Returns the
+     * default-constructed DecodeInfo if there was a decoding error for the
+     * instruction.
+     */
+    DecodeInfo *
+    get_decode_info(app_pc pc)
+    {
+        auto it = decode_cache_.find(pc);
+        if (it == decode_cache_.end()) {
+            return nullptr;
+        }
+        return &it->second;
+    }
+    /**
+     * Adds decode info for the given \p memref_instr if it is not yet recorded.
+     */
+    void
+    add_decode_info(const dynamorio::drmemtrace::_memref_instr_t &memref_instr)
+    {
+        const app_pc trace_pc = reinterpret_cast<app_pc>(memref_instr.addr);
+        if (memref_instr.encoding_is_new) {
+            decode_cache_.erase(trace_pc);
+        }
+        if (decode_cache_.find(trace_pc) != decode_cache_.end()) {
+            return;
+        }
+        decode_cache_[trace_pc] = DecodeInfo();
+        instr_t *instr = nullptr;
+        instr_noalloc_t noalloc;
+        if (persist_decoded_instrs_) {
+            instr = instr_create(dcontext_);
+        } else {
+            instr_noalloc_init(dcontext_, &noalloc);
+            instr = instr_from_noalloc(&noalloc);
+        }
+        app_pc next_pc = decode_from_copy(
+            dcontext_, const_cast<app_pc>(memref_instr.encoding), trace_pc, instr);
```

Comment:
Comment somewhere that the trace must have OFFLINE_FILE_VERSION_ENCODINGS set in order to use this library.  Maybe even provide a check_filetype() library function to make that easy.

Reply:
Added a comment. I feel check_filetype wouldn't help much unless we make it a fatal error to call any decode cache API without calling check_filetype (if one remembers to call check_filetype, they are likely already aware of the limitation). Was that what you suggested?

Reply:
What will happen if someone uses this library on an x86 trace with no encodings (such as one of our public v1 traces) without checking anything?  That seems not at all unlikely for someone external.  If the memref struct is zero-initialized they'll get back valid decodings for every single instruction claiming it's `add [rax], al` or sthg. I think currently the analyzer just declares RecordType and memref_t has no initialization of all the fields: so it will be garbage, which will probably still decode to some valid instruction much of the time.

I guess checking whether the user called the check routine is a possibility; I was thinking of that being optional and more about helping newbies easily check whether this library can be used on a trace.

Could we check the instr size field and have a fatal error if the instr size doesn't match the decoded size?  That should give us invalid decodings most of the time with a no-encoding trace and seems a good sanity check in any case.

Reply:
> Could we check the instr size field and have a fatal error if the instr size doesn't match the decoded size?

But those two fields are indeed different for the synthetic ISA instrs (right?).

Reply:
Oh right, that is true (right @edeiana ?)

Also: isn't the filetype available at init time now?  So could we have an init time check?

Reply:
An init-time check still requires the user to call check_filetype, which isn't useful unless we make absence of such a call fatal.

See other comment thread: a better way would be to check encoding_is_new in add_decode_info calls, which will never be true for a trace without embedded-encodings.

Reply:
> Also: isn't the filetype available at init time now? 

Doesn't seem to be the case for `-infile` where we set `read_inputs_in_init = false`.

Reply:
-infile does not set that field to false.  I think you mean online analysis, which sets that to false.

Reply:
Oh right, thanks for correcting.

Reply:
That's still a reason for tools to not do the init() in their initialize_stream(), but during memref processing.

Reply:
Though do we support modules in online analysis at all? Maybe not.

Reply:
Added the filetype check in init()